run-command.c

```diff
@@ -795,6 +796,11 @@ int start_command(struct child_process *cmd)
 		if (cmd->dir && chdir(cmd->dir))
 			child_die(CHILD_ERR_CHDIR);
 
+		if (cmd->close_parent_file_descriptors) {
+			for (fd = 3; fd < 256; ++fd)
+				set_cloexec(fd);
```

Comment:
Careful here. Some of these descriptors might correspond to the end points retrieved by the `pipe()` call when `in`, `out` or `err` in the `child_process` are set to `-1`. I think you will need to test for those, like so:

```c
if (fd != cmd->in && fd != cmd->out && fd != cmd->err)
    set_cloexec(fd);
```

With this change, you should even be able to start at `fd = 0` in that loop.

Reply:
When `in`, `out`, or `err` are set to `-1`, the parent process calls `pipe()` and saves its end of the pipe in the `child_process` corresponding field. 

In the child process it calls `child_dup2` to hook up its end of the pipe to `0`, `1`, or `2`.  It then closes the pipe descriptors with a call to `child_close_pair` or `child_close` when it is not `-1`.

So the child process should always have `0`, `1`, and `2` and the pipe if used is connected to one of those.